File: packages/dart/lib/src/kiwoom_client.dart
import 'dart:io';
import 'client.dart';
import 'generated/api_paths.dart';

class SegmentClient {
  const SegmentClient(this._segment);
  final String _segment;

  Future<Map<String, dynamic>> call(
    String apiId, [
    Map<String, dynamic> args = const {},
  ]) {
    final realPath = kApiPaths[apiId];
    if (realPath == null) throw ArgumentError('Unknown api_id: $apiId');
    if (kApiSegments[apiId] != _segment) {
      throw ArgumentError(
          "api_id $apiId belongs to segment '${kApiSegments[apiId]}', not '$_segment'");
    }
    return callKiwoom(realPath, apiId, args);
  }
}

class KiwoomClient {
  KiwoomClient({
    String? appKey,
    String? appSecret,
    String? env,
    String? accessToken,
  }) {
    if (appKey != null) _setEnv('KIWOOM_APP_KEY', appKey);
    if (appSecret != null) _setEnv('KIWOOM_APP_SECRET', appSecret);
    if (env != null) _setEnv('KIWOOM_ENV', env);
    if (accessToken != null) _setEnv('KIWOOM_ACCESS_TOKEN', accessToken);
  }

  void _setEnv(String key, String value) {
    // Platform.environment is read-only; use dart:io Process.run workaround
    // For env injection, users should set via system env or MCP config.
    // This constructor stores values for potential future use.
    // Actual env reading happens in auth.dart via Platform.environment.
    _overrides[key] = value;
  }

  final Map<String, String> _overrides = {};

  final acnt = const SegmentClient('acnt');
  final chart = const SegmentClient('chart');
  final crdordr = const SegmentClient('crdordr');
  final elw = const SegmentClient('elw');
  final etf = const SegmentClient('etf');
  final frgnistt = const SegmentClient('frgnistt');
  final mrkcond = const SegmentClient('mrkcond');
  final ordr = const SegmentClient('ordr');
  final rkinfo = const SegmentClient('rkinfo');
  final sect = const SegmentClient('sect');
  final shsa = const SegmentClient('shsa');
  final slb = const SegmentClient('slb');
  final stkinfo = const SegmentClient('stkinfo');
  final thme = const SegmentClient('thme');

  Future<Map<String, dynamic>> call(
    String apiId, [
    Map<String, dynamic> args = const {},
  ]) {
    final realPath = kApiPaths[apiId];
    if (realPath == null) throw ArgumentError('Unknown api_id: $apiId');
    return callKiwoom(realPath, apiId, args);
  }

  Map<String, String> get envOverrides => Map.unmodifiable(_overrides);
}

File: packages/typescript/src/KiwoomClient.ts
import { callKiwoom, KiwoomResponse } from "./client.js";
import { API_PATHS, API_SEGMENTS } from "./generated/api_paths.js";

export class SegmentClient {
  constructor(private readonly _segment: string) {}

  call(apiId: string, args: Record<string, unknown> = {}): Promise<KiwoomResponse> {
    const realPath = API_PATHS[apiId];
    if (!realPath) throw new Error(`Unknown api_id: ${apiId}`);
    if (API_SEGMENTS[apiId] !== this._segment) {
      throw new Error(
        `api_id ${apiId} belongs to segment '${API_SEGMENTS[apiId]}', not '${this._segment}'`
      );
    }
    return callKiwoom(realPath, apiId, args);
  }
}

export interface KiwoomClientConfig {
  appKey?: string;
  appSecret?: string;
  env?: "live" | "mock";
  accessToken?: string;
}

export class KiwoomClient {
  readonly acnt = new SegmentClient("acnt");
  readonly chart = new SegmentClient("chart");
  readonly crdordr = new SegmentClient("crdordr");
  readonly elw = new SegmentClient("elw");
  readonly etf = new SegmentClient("etf");
  readonly frgnistt = new SegmentClient("frgnistt");
  readonly mrkcond = new SegmentClient("mrkcond");
  readonly ordr = new SegmentClient("ordr");
  readonly rkinfo = new SegmentClient("rkinfo");
  readonly sect = new SegmentClient("sect");
  readonly shsa = new SegmentClient("shsa");
  readonly slb = new SegmentClient("slb");
  readonly stkinfo = new SegmentClient("stkinfo");
  readonly thme = new SegmentClient("thme");

  constructor(config: KiwoomClientConfig = {}) {
    if (config.appKey) process.env.KIWOOM_APP_KEY = config.appKey;
    if (config.appSecret) process.env.KIWOOM_APP_SECRET = config.appSecret;
    if (config.env) process.env.KIWOOM_ENV = config.env;
    if (config.accessToken) process.env.KIWOOM_ACCESS_TOKEN = config.accessToken;
  }

  call(apiId: string, args: Record<string, unknown> = {}): Promise<KiwoomResponse> {
    const realPath = API_PATHS[apiId];
    if (!realPath) throw new Error(`Unknown api_id: ${apiId}`);
    return callKiwoom(realPath, apiId, args);
  }
}
